Fix form field joomuser popup

Erftralle opened the pull request:
Due to recent changes of the user form field in Joomla 3.5 the user resp owner selection via popup with the joomuser form field is not working any more. You can check this by trying to assign an owner to a category in the backend category view.

This pull request fixes that.

Erftralle committed:
Fix broken user selection via popup in JFormFieldJoomUser

Erftralle committed:
Typo

Hackwar commented:
Works fine for me

Hackwar commented:
I've updated an installation with the latest master and this PR and while this improves a lot of things, there is still an issue in the filter. When opening the search tools and selecting a filter, it will automatically set the owner filter to 0 instead of ''. Since I need to filter both for owned and unowned images and also unrelated to the owner, we would need a tri-value boolean here. :wink:  Do you have an idea how to fix that?

Erftralle committed:
Correct owner filter value

Erftralle commented:
If the system has less than 250 users the owner filter is presented as select box. This one should work without problems.
If the system has more than 250 users the modal popup is used.
The last commit should fix the problem. Not a nice solution, but I had no other ideas.

coolman01 commented:
Hi
Obviously I am being quite dim, but I don't understand what I have to do to implement this fix in the website. What do I download and from where and what do I have to to install it?
Grateful for help.

Chraneco commented:
Thank you very much!

@coolman01 This is now merged and will be included in the next release. If you need the fix now you can download the zip archive of this branch: https://github.com/Erftralle/JoomGallery/tree/fix-form-field-joomuser-popup

coolman01 commented:
@Chraneco 
Great stuff!
Thanks very much.

Chraneco commented on administrator/components/com_joomgallery/models/images.php:
```diff
@@ -282,6 +282,13 @@ protected function populateState($ordering = 'a.ordering', $direction = 'asc')
+        if($name == 'owner' && $value == 0)
+        {
+          $value = '';
+          JFactory::getApplication()->setUserState('joom.images.filter.' . $name, $value);
```
If you are logged into the backend with a fresh session it is not possible to switch to a different page in the image manager with the first attempt. When the page reloads you are still on the first page but every subsequent click on the pagination works as expected.
I could trace this problem back to the line above. If that line is commented out the pagination works even on the first attempt.
What changed in Joomla! 3.5 which makes this line necessary?

Erftralle commented:
> What changed in Joomla! 3.5 which makes this line necessary?

Before Joomla! 3.5 there has been no layout file for the user modal. In that new layout file a owner filter value of '' is casted to an int of 0 (see https://github.com/joomla/joomla-cms/blob/staging/layouts/joomla/form/field/user.php#L82). But all the Joomla! filter stuff needs a value of '' for an empty filter.

The reason why I added these lines was the post of @Hackwar above:

> I've updated an installation with the latest master and this PR and while this improves a lot of things, there is still an issue in the filter. When opening the search tools and selecting a filter, it will automatically set the owner filter to 0 instead of ''. Since I need to filter both for owned and unowned images and also unrelated to the owner, we would need a tri-value boolean here. :wink: Do you have an idea how to fix that?

I swear that I was able to reproduce that. So I repeated my tests and today I was not able to reproduce it again. This is quite obvious if we have a look in the `_buildQuery()` of the images model. It does not matter if the owner value is 0 or ''. The filter is not applied in both cases.

What I can confirm is that these lines are causing pagination issues. So I think we should remove them again :wink:.

If the select box is used as filter box all is ok for me after removing the code again. 

But I found some other glitches when the modal popup is used:
If you are logged into the backend with a fresh session the first change to another page works fine. But you will see an open filter bar which should be closed. If you press reset the filter bar is closed again and page one is shown. If you now want to change the page again you now need two attempts for it.

Do you have an idea how two solve these two bugs then? 

What do you think of adding the following code in `getUserStateFromRequest()` directly behind

`$new_state = JRequest::getVar($request, null, 'default', $type);`


`    if(     !is_null($new_state) && isset($new_state['owner'])`
`        && ($new_state['owner'] === 0 || $new_state['owner'] === '0'))`
`    {`
`      $new_state['owner'] = '';`
`    }`

Chraneco commented:
I see, thanks for the explanations!
What about just adding the condition `$this->state->{$filterName} !== '0'` when the active filters are detected in method `JoomGalleryModelImages::getActiveFilters`?

It seems to solve these bugs for me, it would just not be possible anymore to filter for the string '0' in any filter, but I'm not sure whether that's needed.

Erftralle commented:
Do you mean changing the condition to 
`        if (property_exists($this->state, $filterName) && $this->state->{$filterName} !== '0' && (!empty($this->state->{$filterName}) || is_numeric($this->state->{$filterName}))) ` ?

In my tests it does not solve the second bug mentioned above. After resetting the filters you still need two times to change to a different page.

> it would just not be possible anymore to filter for the string '0' in any filter, but I'm not sure whether that's needed.

I don't know if it's really needed but maybe someone wants to search for '0' in the image title :wink: .

Chraneco commented:
Oh, sorry I forgot to test that. The problem there seems to be that the JavaScript that is executed when hitting the clear button is setting all fields to an empty string, even the one that Joomla previously forces to an integer. This is really inconsistent, but unfortunately that's not our code.

I wanted to avoid adding code concerning a specific field like the `owner` field into this general method `getUserStateFromRequest`, but it looks like your suggestion is the only way to fix that issue.

Erftralle commented:
> I wanted to avoid adding code concerning a specific field like the `owner` field into this general method `getUserStateFromRequest`, but it looks like your suggestion is the only way to fix that issue.

I agree, that's not a nice solution but I did not find a better one either. I created #97.

Chraneco commented:
Thanks for the pull request! I merged it.
Unfortunately, I found another issue while testing which occurs with our without this last change: It is not possible to filter for '- No User -'. Since this is probably happening because '- No User -' and ' No Filter' is identified with the same values I'm not sure how to solve this problem.

Erftralle commented:
> Unfortunately, I found another issue while testing which occurs with our without this last change: It is not possible to filter for '- No User -'. Since this is probably happening because '- No User -' and ' No Filter' is identified with the same values I'm not sure how to solve this problem.

As far as I know it has never been possible before to filter '- No User -' with the owner filter. For that you may use the type filter by selecting 'Only Admin Uploaded Images' which filters all images having the owner value 0.
